File: ml/ensemble.py
"""Ensemble methods: RandomForest."""

from __future__ import annotations

import numpy as np
from .tree import DecisionTree


class RandomForest:
    def __init__(
        self,
        n_estimators: int = 10,
        max_depth: int = 10,
        min_samples_split: int = 2,
    ) -> None:
        self.n_estimators: int = n_estimators
        self.max_depth: int = max_depth
        self.min_samples_split: int = min_samples_split
        self.trees: list[DecisionTree] = []
        self.is_classification: bool = True
        self.n_classes: int = 0

    def fit(self, X: np.ndarray, y: np.ndarray) -> None:
        self.is_classification = len(np.unique(y)) < len(y) / 2
        self.n_classes = len(np.unique(y))
        self.trees = []
        n_samples = X.shape[0]

        for _ in range(self.n_estimators):
            indices = np.random.choice(n_samples, n_samples, replace=True)
            X_boot = X[indices]
            y_boot = y[indices]
            tree = DecisionTree(max_depth=self.max_depth, min_samples_split=self.min_samples_split)
            tree.fit(X_boot, y_boot)
            self.trees.append(tree)

    def predict(self, X: np.ndarray) -> np.ndarray:
        predictions = np.array([tree.predict(X) for tree in self.trees])
        if self.is_classification:
            majority_votes = np.apply_along_axis(lambda x: np.bincount(x.astype(int)).argmax(), 0, predictions)
            return majority_votes
        return np.mean(predictions, axis=0)

File: ml/ensemble.rs
//! ml/ensemble.rs
//! Ensemble methods: RandomForest.

use rand::{Rng, SeedableRng};
use super::tree::DecisionTree;

pub struct RandomForest {
    pub n_estimators: usize,
    pub max_depth: usize,
    pub min_samples_split: usize,
    pub trees: Vec<DecisionTree>,
    pub is_classification: bool,
    pub n_classes: usize,
}

impl RandomForest {
    pub fn new(n_estimators: usize, max_depth: usize, min_samples_split: usize) -> Self {
        RandomForest {
            n_estimators,
            max_depth,
            min_samples_split,
            trees: Vec::new(),
            is_classification: true,
            n_classes: 0,
        }
    }

    pub fn fit(&mut self, X: &[Vec<f64>], y: &[f64]) {
        let mut unique: Vec<f64> = y.to_vec();
        unique.sort_by(|a, b| a.partial_cmp(b).unwrap());
        unique.dedup_by(|a, b| (*a - *b).abs() < 1e-9);
        
        self.is_classification = unique.len() < y.len() / 2;
        self.n_classes = unique.len();
        self.trees = Vec::new();
        
        let n_samples = X.len();
        let mut rng = rand::rngs::StdRng::from_entropy();
        
        for _ in 0..self.n_estimators {
            let mut indices = Vec::with_capacity(n_samples);
            for _ in 0..n_samples {
                indices.push(rng.gen_range(0..n_samples));
            }
            
            let mut X_boot = Vec::with_capacity(n_samples);
            let mut y_boot = Vec::with_capacity(n_samples);
            
            for &idx in &indices {
                X_boot.push(X[idx].clone());
                y_boot.push(y[idx]);
            }
            
            let mut tree = DecisionTree::new(self.max_depth, self.min_samples_split);
            tree.fit(&X_boot, &y_boot);
            self.trees.push(tree);
        }
    }

    pub fn predict(&self, X: &[Vec<f64>]) -> Vec<f64> {
        let predictions: Vec<Vec<f64>> = self.trees.iter()
            .map(|tree| tree.predict(X))
            .collect();
        
        if self.is_classification {
            let mut result = Vec::with_capacity(X.len());
            for j in 0..X.len() {
                let mut counts = std::collections::HashMap::new();
                for tree in &self.trees {
                    let pred = predictions.iter().nth(tree as *const _ as usize).and_then(|p| p.get(j));
                    if let Some(&p) = pred {
                        *counts.entry(p as i32).or_insert(0) += 1;
                    }
                }
                let majority = counts.iter().max_by_key(|(_, &c)| c).map(|(&k, _)| k as f64).unwrap_or(0.0);
                result.push(majority);
            }
            result
        } else {
            let n_trees = self.trees.len();
            let mut result = Vec::with_capacity(X.len());
            for j in 0..X.len() {
                let sum: f64 = self.trees.iter()
                    .map(|tree| tree.predict(&[X[j].to_vec()])[0])
                    .sum();
                result.push(sum / n_trees as f64);
            }
            result
        }
    }
}